bucket-handlers.go

```diff
@@ -356,65 +385,51 @@ func (api storageAPI) PutBucketHandler(w http.ResponseWriter, r *http.Request) {
 	vars := mux.Vars(r)
 	bucket := vars["bucket"]
 
-	if isRequestRequiresACLCheck(r) {
-		writeErrorResponse(w, r, AccessDenied, r.URL.Path)
-		return
-	}
-
-	// read from 'x-amz-acl'
-	aclType := getACLType(r)
-	if aclType == unsupportedACLType {
-		writeErrorResponse(w, r, NotImplemented, r.URL.Path)
-		return
-	}
-
-	// if body of request is non-nil then check for validity of Content-Length
-	if r.Body != nil {
-		/// if Content-Length is unknown/missing, deny the request
-		if r.ContentLength == -1 && !contains(r.TransferEncoding, "chunked") {
-			writeErrorResponse(w, r, MissingContentLength, r.URL.Path)
-			return
-		}
-	}
-
 	// Set http request for signature.
 	auth := api.Signature.SetHTTPRequestToVerify(r)
-
-	if isRequestPresignedSignatureV4(r) {
-		ok, err := auth.DoesPresignedSignatureMatch()
-		if err != nil {
-			errorIf(err.Trace(r.URL.String()), "Presigned signature verification failed.", nil)
-			writeErrorResponse(w, r, SignatureDoesNotMatch, r.URL.Path)
+	if isRequestAnonymous(r) {
+		// http://docs.aws.amazon.com/AmazonS3/latest/dev/mpuAndPermissions.html
+		if isAllowed, s3Error := checkAccess("s3:CreateBucket", bucket, r.URL); !isAllowed {
+			writeErrorResponse(w, r, s3Error, r.URL.Path)
 			return
 		}
-		if !ok {
-			writeErrorResponse(w, r, SignatureDoesNotMatch, r.URL.Path)
-			return
-		}
-	} else if isRequestSignatureV4(r) {
-		// Verify signature for the incoming body if any.
-		locationBytes, e := ioutil.ReadAll(r.Body)
-		if e != nil {
-			errorIf(probe.NewError(e), "MakeBucket failed.", nil)
-			writeErrorResponse(w, r, InternalError, r.URL.Path)
-			return
-		}
-		sh := sha256.New()
-		sh.Write(locationBytes)
-		ok, err := auth.DoesSignatureMatch(hex.EncodeToString(sh.Sum(nil)))
-		if err != nil {
-			errorIf(err.Trace(), "MakeBucket failed.", nil)
-			writeErrorResponse(w, r, InternalError, r.URL.Path)
-			return
-		}
-		if !ok {
-			writeErrorResponse(w, r, SignatureDoesNotMatch, r.URL.Path)
-			return
+	} else {
+		if isRequestPresignedSignatureV4(r) {
+			ok, err := auth.DoesPresignedSignatureMatch()
+			if err != nil {
+				errorIf(err.Trace(r.URL.String()), "Presigned signature verification failed.", nil)
+				writeErrorResponse(w, r, SignatureDoesNotMatch, r.URL.Path)
+				return
+			}
+			if !ok {
+				writeErrorResponse(w, r, SignatureDoesNotMatch, r.URL.Path)
+				return
+			}
+		} else if isRequestSignatureV4(r) {
+			// Verify signature for the incoming body if any.
```

Comment:
In many places, "else if" cases have no default case. Such ignored errors are hard to trace.